File: othello/display.py
"""Terminal rendering for boards, moves, and scores (ANSI colour)."""
from __future__ import annotations
from typing import Final, Callable, TypeGuard, cast

from othello.core import (
    Board, Bitmap, Move, Player, Score, PASS, BLACK,
    format_square, format_move,
)

CELL_WIDTH: Final[int] = 2

RESET = "\033[0m"

BLACK_FG = "\033[90m"
WHITE_FG = "\033[97m"      # bright white

BLACK_BG = "\033[97;100m"
WHITE_BG = "\033[30;107m"

BLACK_DOT_FG = "\033[38;5;240m"
WHITE_DOT_FG = "\033[38;5;255m"

EMPTY_CELL = "\033[2;38;5;240m·\033[0m "

BLACK_MARK  = f"{BLACK_DOT_FG}⬤{RESET} "
WHITE_MARK  = f"{WHITE_DOT_FG}⬤{RESET} "

BLACK_LEGAL = f"{BLACK_FG}◌ {RESET}"
WHITE_LEGAL = f"{WHITE_FG}◌ {RESET}"

GREEN_FG = "\033[38;5;46m"
LEGAL_DOT = f"{GREEN_FG}•{RESET} "

RED_BG = "\033[48;5;52m"

BLACK_LAST_MARK = f"{RED_BG}{BLACK_DOT_FG}⬤{RESET} "
WHITE_LAST_MARK = f"{RED_BG}{WHITE_DOT_FG}⬤{RESET} "

type CellRenderer = str | Callable[[int], str]

def is_renderer(x: CellRenderer) -> TypeGuard[Callable[[int], str]]:
    return not isinstance(x, str)

type BitmapLayer = tuple[Bitmap, CellRenderer]


# Rank/file legal-move indicators — replaced by LEGAL_DOT (green dot).
# def black_legal(sq: int) -> str:
#     return f"{BLACK_BG}{format_square(sq):<{CELL_WIDTH}}{RESET}"
#
# def white_legal(sq: int) -> str:
#     return f"{WHITE_BG}{format_square(sq):<{CELL_WIDTH}}{RESET}"


def last_move_renderer(board: Board) -> Callable[[int], str]:
    def render(sq: int) -> str:
        bit = 1 << sq
        if board.black & bit:
            return BLACK_LAST_MARK
        if board.white & bit:
            return WHITE_LAST_MARK
        return f"\033[103m{format_square(sq):<{CELL_WIDTH}}\033[0m"
    return render

def format_board(board: Board, show_valid_moves=True, last_move: Move | None = None) -> str:
    layers: list[BitmapLayer] = []
    if last_move is not None and last_move != PASS:
        if last_move & (last_move - 1):
            raise ValueError(f"last_move is not one-hot: {last_move:#x}")
        layers.append((last_move, last_move_renderer(board)))
    layers.extend([
        (board.black, BLACK_MARK),
        (board.white, WHITE_MARK),
    ])
    if show_valid_moves:
        layers.append((board.actions(), LEGAL_DOT))
    return _format_board(*layers)


def _format_board(*layers: BitmapLayer) -> str:
    out: list[str] = []
    for r in range(7, -1, -1):
        row: list[str] = []
        for c in range(8):
            sq = r * 8 + c
            bit = 1 << sq
            cell: str = EMPTY_CELL
            for bitmap, renderer in layers:
                if bitmap & bit:
                    if is_renderer(renderer):
                        cell = renderer(sq)
                    else:
                        cell = cast(str, renderer)
                    break
            row.append(cell)
        out.append(f"{r + 1:>2}  {' '.join(row)}")
    out.append("    " + " ".join(f"{f:<{CELL_WIDTH}}" for f in "ABCDEFGH"))
    return "\n".join(out)


def player_name(player: Player) -> str:
    return "black" if player == BLACK else "white"

def move_name(move: Move) -> str:
    return "PASS" if move == PASS else format_move(move)

def format_score(score: Score) -> str:
    # Black-centered margin, sign kept, tagged by who it favours: B:+6 = black
    # ahead by 6, W:-4 = white ahead by 4 (black-centered -4), T:0 = tie.
    if score > 0:
        return f"B:{score:+d}"
    if score < 0:
        return f"W:{score:+d}"
    return "T:0"

File: rust/src/display.rs
//! Terminal rendering for boards, moves, and scores (ANSI colour), port of
//! `display.py`.

use crate::core::{format_move, format_square, Board, Move, Player, Score, BLACK, PASS};

pub const RESET: &str = "\x1b[0m";
pub const BLACK_DOT_FG: &str = "\x1b[38;5;240m";

const BLACK_BG: &str = "\x1b[97;100m";
const WHITE_BG: &str = "\x1b[30;107m";

const EMPTY_CELL: &str = "\x1b[2;38;5;240m\u{00b7}\x1b[0m ";
const BLACK_MARK: &str = "\x1b[38;5;240m\u{2b24}\x1b[0m ";
const WHITE_MARK: &str = "\x1b[38;5;255m\u{2b24}\x1b[0m ";
const BLACK_LAST_MARK: &str = "\x1b[48;5;52m\x1b[38;5;240m\u{2b24}\x1b[0m ";
const WHITE_LAST_MARK: &str = "\x1b[48;5;52m\x1b[38;5;255m\u{2b24}\x1b[0m ";

pub fn format_board(board: &Board, show_valid_moves: bool, last_move: Option<Move>) -> String {
    let actions = if show_valid_moves { board.actions() } else { 0 };
    let last = match last_move {
        Some(m) if m != PASS => Some(m),
        _ => None,
    };

    let mut out = String::new();
    for r in (0..8u32).rev() {
        let mut row: Vec<String> = Vec::with_capacity(8);
        for c in 0..8u32 {
            let sq = r * 8 + c;
            let bit = 1u64 << sq;
            let cell = if last.is_some_and(|m| m & bit != 0) {
                if board.black & bit != 0 {
                    BLACK_LAST_MARK.to_string()
                } else if board.white & bit != 0 {
                    WHITE_LAST_MARK.to_string()
                } else {
                    format!("\x1b[103m{:<2}\x1b[0m", format_square(sq))
                }
            } else if board.black & bit != 0 {
                BLACK_MARK.to_string()
            } else if board.white & bit != 0 {
                WHITE_MARK.to_string()
            } else if actions & bit != 0 {
                let bg = if board.to_move == BLACK {
                    BLACK_BG
                } else {
                    WHITE_BG
                };
                format!("{bg}{:<2}{RESET}", format_square(sq))
            } else {
                EMPTY_CELL.to_string()
            };
            row.push(cell);
        }
        out.push_str(&format!("{:>2}  {}\n", r + 1, row.join(" ")));
    }
    let files: Vec<String> = "ABCDEFGH".chars().map(|f| format!("{f:<2}")).collect();
    out.push_str(&format!("    {}", files.join(" ")));
    out
}

pub fn player_name(player: Player) -> &'static str {
    if player == BLACK {
        "black"
    } else {
        "white"
    }
}

pub fn move_name(mv: Move) -> String {
    if mv == PASS {
        "PASS".to_string()
    } else {
        format_move(mv)
    }
}

pub fn format_score(score: Score) -> String {
    // Black-centred margin: B:+6 = black ahead by 6, W:-4 = white ahead by 4.
    if score > 0 {
        format!("B:{score:+}")
    } else if score < 0 {
        format!("W:{score:+}")
    } else {
        "T:0".to_string()
    }
}
